icu4c/source/common/rbbinode.cpp

```diff
@@ -239,7 +239,14 @@ RBBINode *RBBINode::cloneTree() {
 //                      nested references are handled by cloneTree(), not here.
 //
 //-------------------------------------------------------------------------
-RBBINode *RBBINode::flattenVariables() {
+constexpr int kRecursiveDepthLimit = 3500;
+RBBINode *RBBINode::flattenVariables(UErrorCode& status, int depth) {
+    // If the depth of the stack is too deep, we return U_BRK_INTERNAL_ERROR
+    // to avoid stack overflow crash.
+    if (depth > kRecursiveDepthLimit) {
```

Comment:
Need to check for an incoming error first, and not overwrite an existing error with a new one.